pkg/middlewares/tcp/streamcompress/stream_compress_test.go

```diff
@@ -0,0 +1,216 @@
+	require.NoError(t, err)
+
+	compressorConfig := dynamic.TCPStreamCompress{
+		Algorithm: "zstd",
+		Level:     "best",
+		Upstream:  false,
+	}
+
+	compressor, err := New(context.Background(), decompressor, compressorConfig, "traefikTest")
+	require.NoError(t, err)
+
+	server, client := net.Pipe()
+
+	go func() {
+		compressor.ServeTCP(&contextWriteCloser{client, addr{"10.10.10.10"}})
+	}()
+
+	read, err := io.ReadAll(server)
+	require.NoError(t, err)
+
+	assert.Equal(t, message, string(read))
+
+	err = server.Close()
+	require.NoError(t, err)
+}
+
+func TestStreamCompress_ServeTCPCompression(t *testing.T) {
+	next := tcp.HandlerFunc(func(conn tcp.WriteCloser) {
+		write, err := conn.Write([]byte(message))
+		require.NoError(t, err)
+		assert.Equal(t, len(message), write)
+
+		err = conn.Close()
+		require.NoError(t, err)
+	})
+
+	decompressorConfig := dynamic.TCPStreamCompress{
+		Algorithm: "zstd",
+		Level:     "best",
+		Upstream:  true,
+	}
+
+	decompressor, err := New(context.Background(), next, decompressorConfig, "traefikTest3")
+	require.NoError(t, err)
+
+	server, client := net.Pipe()
+
+	go func() {
+		decompressor.ServeTCP(&contextWriteCloser{client, addr{"10.10.10.10"}})
+	}()
+
+	serverDecompressor, err := zstd.NewReader(server)
+	require.NoError(t, err)
+
+	read, err := io.ReadAll(serverDecompressor)
+	require.NoError(t, err)
+
+	assert.Equal(t, message, string(read))
+
+	err = server.Close()
+	require.NoError(t, err)
+
+	serverDecompressor.Close()
+}
+
+func TestStreamCompress_ServeTCPDecompression(t *testing.T) {
+	next := tcp.HandlerFunc(func(conn tcp.WriteCloser) {
+		compressor, err := zstd.NewWriter(conn)
+		require.NoError(t, err)
+
+		write, err := compressor.Write([]byte(message))
+		require.NoError(t, err)
+		assert.Equal(t, len(message), write)
+
+		err = compressor.Flush()
+		require.NoError(t, err)
+
+		time.Sleep(100 * time.Millisecond)
+
+		err = conn.Close()
+		require.NoError(t, err)
+	})
+
+	decompressorConfig := dynamic.TCPStreamCompress{
+		Algorithm: "zstd",
+		Level:     "best",
+		Upstream:  false,
+	}
+
+	decompressor, err := New(context.Background(), next, decompressorConfig, "traefikTest3")
+	require.NoError(t, err)
+
+	server, client := net.Pipe()
+
+	go func() {
+		decompressor.ServeTCP(&contextWriteCloser{client, addr{"10.10.10.10"}})
+	}()
+
+	read, err := io.ReadAll(server)
+	require.NoError(t, err)
+
+	assert.Equal(t, message, string(read))
+
+	err = server.Close()
+	require.NoError(t, err)
+}
+
+type contextWriteCloser struct {
+	net.Conn
+	addr
+}
+
+type addr struct {
+	remoteAddr string
+}
+
+func (a addr) Network() string {
+	panic("implement me")
+}
+
+func (a addr) String() string {
+	return a.remoteAddr
+}
+
+func (c contextWriteCloser) CloseWrite() error {
+	panic("implement me")
+}
+
+func (c contextWriteCloser) RemoteAddr() net.Addr {
+	return c.addr
+}
+
+func (c contextWriteCloser) Context() context.Context {
+	return context.Background()
+}
```

Comment:
Could you please add some automated benchmarking functions here also? 2 examples come to mind:

1. spin up configurable number (eg default of 10000) independent connections to a server, send and receive 50kb of data, measure overall golang memory usage and time
2. wrap-nest configurable (eg 100) decoder/encoders in a pipeline on single connection, send data through and measure mem usage and time. I believe this is the test you've been doing and automating locally anyway?

Reply:
ok thanks can you put results of this in the ticket? perhaps a few to see how (1) scales with number of clients? (& mem usage..)